### PR TITLE
Expose new mount points structs in inspect.

### PR DESCRIPTION
**Pull request by calavera**
Docker Inspect presents the volume information in two different hashes. There are a couple of issues with this approach:
1. It's hard to parse, because if you want to have the full information for a volume you have to parse the output twice, one to look at the sources information and another time to look at the permissions information.
2. It cannot display any more information besides what's in it now, because it's designed with a map of strings and a map of boolean. The amount of information for a mount point is growing with named volumes, external drivers and so on, and we need to show this information.

The new approach exposes the full mount point structure as a single element in the output. So you only need to parse a single field to get the information for a specific volume, and you get everything related to that volume.

**Before this patch:**

``` json
{
  "Volumes": {"/vol1": "/vol1"},
  "VolumesRW": {"/vol1": true}
}
```

**After this patch:**

``` json
{
  "Mounts": [
     {
        "Source": "/vol1",
        "Destination": "/vol1",
        "Mode": "rw,Z",
        "RW": true
     }
  ]
}
```

TODO:
- [x] Tests
- [x] Docs

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

File: api/server/server.go [hunk 1]
```diff
@@ -1142,7 +1142,7 @@ func (s *Server) getContainersByName(version version.Version, w http.ResponseWri
 	}
 
 	if version.LessThan("1.19") {
```
Comment (thaJeztah) [inline api/server/server.go]: Should this be less than 1.20?

Comment (calavera) [inline api/server/server.go]: Yes, fixed!

File: docs/man/docker-inspect.1.md [hunk 2]
```diff
@@ -158,8 +158,13 @@ To get information on a container use its ID or instance name:
             "Running": true,
             "StartedAt": "2014-04-04T21:33:54.16259207Z",
         },
-        "Volumes": {},
-        "VolumesRW": {},
+        "MountPoints": {
+          "/data": {
+            "Source": "/data",
+            "Destination": "/data",
+            "RW" false
```
Comment (vdemeester) [inline docs/man/docker-inspect.1.md]: Question on this one :
- Seems like there is a type (missing a ``:``)
- the ``:ro`` option is mapped to ``"RW": false`` and ``:rw`` to ``"RW": true``, right ? Then what about the ``z`` option, and the composition of them ``rw,z`` ? (for reference : https://github.com/docker/docker/pull/12572 )

Comment (thaJeztah) [inline docs/man/docker-inspect.1.md]: There were some other formatting issues in the docs as well, but better hold that until we're in "docs review", otherwise it gets confusing. :)

Comment (calavera) [inline docs/man/docker-inspect.1.md]: @vdemeester I've added the `Relabel` field, thanks for checking!

**Comment by vbatts** [comment]
cc @rhatdan @rhvgoyal

**Comment by rhatdan** [comment]
Should we include the mount type, in case my tmpfs stuff ever gets merged?

type "bind"

**Comment by calavera** [comment]
> Should we include the mount type, in case my tmpfs stuff ever gets merged?

I rather not include anything that we don't keep track of internally for now. Although, I can see the type added in a near future.

**Comment by mrunalp** [comment]
:+1:

**Comment by rhatdan** [comment]
Ok, this LGTM also.

File: api/types/types.go [hunk 3]
```diff
@@ -229,13 +227,16 @@ type ContainerJSONBase struct {
...
+	MountPoints map[string]MountPoint
+	Config      *runconfig.Config
 }
 
 // backcompatibility struct along with ContainerConfig
-type ContainerJSONRaw struct {
+type ContainerJSONPre120 struct {
 	*ContainerJSONBase
-	Config *ContainerConfig
+	Volumes   map[string]string
```
Comment (LK4D4) [inline api/types/types.go]: Why we changing old API?

Comment (calavera) [inline api/types/types.go]: we're not changing the old api. Those fields are in `ContainerJSONBase` and need to be present in the old api, that's why I moved them here.

`ContainerJSONBase` doesn't include info about volumes anymore. `ContainerJSONPre120` includes the info for the old api and `ContainerJSON` includes the info for the new api.

File: docs/reference/api/docker_remote_api_v1.20.md [hunk 4]
```diff
@@ -420,8 +424,14 @@ Return low-level information on the container `id`
 			"Running": false,
 			"StartedAt": "2015-01-06T15:47:32.072697474Z"
 		},
-		"Volumes": {},
-		"VolumesRW": {}
+		"MountPoints": {
+				"/data": {
```
Comment (abronan) [inline docs/reference/api/docker_remote_api_v1.20.md]: indentation here ;)

**Comment by abronan** [comment]
LGTM

**Comment by tiborvass** [comment]
I would strongly suggest to have `MountPoints` be an array and not a map. `/vol1` is variable, thus should not be a map key, but a value for a `path` key (or some other name).

```
{
  Mountpoints: [
    {
      "Path": "/vol1",
      ...
    },
    {
      "Path": "/vol2",
      ...
  ],
}
```

**Comment by calavera** [comment]
it was a map to try to facilitate format parsing with a go template but I just realized that go doesn't allow you to do something like `{{ .MountPoints./data }}`:

`panic: template: letter:2: bad character U+002F ‘/'`

I've change it to be an array.

**Comment by calavera** [comment]
@tiborvass let me know what you think about the change.
@abronan can you take a second look? I've changed the map to be an array.

**Comment by cpuguy83** [comment]
@calavera `index .Mountspoints "/data"`

**Comment by cpuguy83** [comment]
@calavera Your test is failing since it's expecting a map.

**Comment by calavera** [comment]
> @calavera index .Mountpoints "/data"

oh that's a good point, I forgot about the index key. @tiborvass any comments? I can still go back to the original version.

**Comment by tiborvass** [comment]
@calavera hmm I personally still think it's better with an array design-wise. Others feel free to chime in /cc @stevvooe

**Comment by cpuguy83** [comment]
I always prefer a map when there is a clear key... But in the end it probably doesn't matter either way, other than passing from go-template.

**Comment by calavera** [comment]
okay, leaving it as it is. The test failure should be fixed now.

**Comment by calavera** [comment]
tests are green :tada:

**Comment by cpuguy83** [comment]
LGTM

**Comment by calavera** [comment]
There was a typo in the docs that should be fixes now

**Comment by abronan** [comment]
still LGTM :+1:

File: api/types/types.go [hunk 5]
```diff
@@ -253,3 +254,12 @@ type ContainerConfig struct {
 	CpuShares  int64
 	Cpuset     string
 }
+
+type MountPoint struct {
```
Comment (stevvooe) [inline api/types/types.go]: Please document struct fields.

**Comment by stevvooe** [comment]
Nit: Just call it `Mounts`.

Typically, boolean fields should be avoided at all. For `RW`, could it just be `perm` with a few string keys? Is this redundant with `Relabel`?

**Comment by calavera** [comment]
@stevvooe we already have boolean fields in the current `VolumesRW`, I want to keep the information that people can extract now and add the information that they don't have. Yes, you can still get the same information from `Relabel`, but you have to parse it yourself and know all the options we manage internally.

**Comment by calavera** [comment]
Renamed `MountPoints` to `Mounts`, I like it. Thanks for the suggestion @stevvooe

**Comment by stevvooe** [comment]
LGTM

**Comment by SvenDowideit** [comment]
Docs LGTM - mind you, if you really wanted a map, you could use a volume UID :)

File: docs/reference/api/docker_remote_api_v1.20.md [hunk 6]
```diff
@@ -140,9 +140,14 @@ Create a container
...
                    "com.example.version": "1.0"
            },
-           "Volumes": {
-                   "/tmp": {}
-           },
+           "Mounts": [
+             {
+               "Source": "/data",
+               "Destination": "/data",
+               "Relabel": "ro,Z",
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.20.md]: this caught my eye; is `ro` supposed to be included in `Relabel`? I'd expected relabel to only contain `Z / z` here.

If this is meant to hold all options, it should probably be renamed to `Options` (and possibly be an array?)

Comment (calavera) [inline docs/reference/api/docker_remote_api_v1.20.md]: We treat this an string internally, I want to leave it that way.

Comment (rhatdan) [inline docs/reference/api/docker_remote_api_v1.20.md]: I believe it should be renamed to mode.  The patch I submitted to do :u changes Relabel to Mode.

Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.20.md]: my point was mainly that (iiuc), `ro`/`rw` is not related to relabeling, so found that a bit odd.

Comment (rhatdan) [inline docs/reference/api/docker_remote_api_v1.20.md]: Right, this is a mode and SELinux will parse it correctly to figure out whether or not it needs to relabel.  Calling it relabel makes you think it is a boolean.

**Comment by calavera** [comment]
Alright, renamed `Relabel` to `Mode`. Back to code review. @thaJeztah, @rhatdan can you take another look?

File: integration-cli/docker_cli_inspect_test.go [hunk 7]
```diff
@@ -265,3 +265,44 @@ func (s *DockerSuite) TestInspectContainerGraphDriver(c *check.C) {
...
+	if m.Source != "/data" {
+		c.Fatalf("Expected source /data, was %s\n", m.Source)
+	}
+
+	if m.Destination != "/data" {
+		c.Fatalf("Expected destination /data, was %s\n", m.Destination)
+	}
+
+	if m.Mode != "ro,z" {
+		c.Fatalf("Expected relabel `ro,z`, was %s\n", m.Mode)
```
Comment (thaJeztah) [inline integration-cli/docker_cli_inspect_test.go]: s/relabel/mode/

File: integration-cli/docker_cli_run_test.go [hunk 8]
```diff
@@ -2302,24 +2302,23 @@ func (s *DockerSuite) TestRunVolumesCleanPaths(c *check.C) {
...
 	c.Assert(err, check.IsNil)
 	if !strings.Contains(out, volumesConfigPath) {
 		c.Fatalf("Volume was not defined for /foo\n%q", out)
 	}
 
-	out, err = inspectFieldMap("dark_helmet", "Volumes", "/bar/")
-	c.Assert(err, check.IsNil)
-	if out != "" {
+	out, err = inspectMountSourceField("dark_helmet", "/bar/")
+	if err != mountNotFound {
```
Comment (thaJeztah) [inline integration-cli/docker_cli_run_test.go]: Should we still be talking about "volumes" here (given that it's all called "mounts")?

File: daemon/inspect.go [hunk 9]
```diff
@@ -20,10 +20,22 @@ func (daemon *Daemon) ContainerInspect(name string) (*types.ContainerJSON, error
...
 
-	return &types.ContainerJSON{base, container.Config}, nil
+	mountPoints := make([]types.MountPoint, 0, len(container.MountPoints))
+	for _, m := range container.MountPoints {
+		mountPoints = append(mountPoints, types.MountPoint{
+			Name:        m.Name,
+			Source:      m.Path(),
+			Destination: m.Destination,
+			Driver:      m.Driver,
+			Mode:        m.Relabel,
```
Comment (thaJeztah) [inline daemon/inspect.go]: Is `m.Relabel` correct here?

Comment (calavera) [inline daemon/inspect.go]: yes

File: integration-cli/docker_cli_run_test.go (hunk 8 above)
Comment (calavera) [inline integration-cli/docker_cli_run_test.go]: yes, those are volumes created from a `VOLUME` directive in a dockerfile.

Comment (thaJeztah) [inline integration-cli/docker_cli_run_test.go]: Ok, thx, got a bit confused :-)

File: api/types/types.go [hunk 10]
```diff
@@ -253,3 +254,13 @@ type ContainerConfig struct {
...
 	Cpuset     string
 }
+
+// MountPoint represents a mount point configuration inside the container.
+type MountPoint struct {
+	Name        string `json:",omitempty"`
+	Source      string
+	Destination string
+	Driver      string `json:",omitempty"`
+	Mode        string
```
Comment (stevvooe) [inline api/types/types.go]: Place a small doc comment for `Mode` that this might be known as `Relabel` internally.

Comment (calavera) [inline api/types/types.go]: done

File: api/types/types.go [hunk 11]
```diff
@@ -235,13 +233,16 @@ type ContainerJSONBase struct {
 
 type ContainerJSON struct {
 	*ContainerJSONBase
```
Comment (stevvooe) [inline api/types/types.go]: I'd recommend against using a pointer embed.

Comment (calavera) [inline api/types/types.go]: that has not changed in this PR, I rather not make changes out of the scope.

**Comment by jfrazelle** [comment]
LGTM, @thaJeztah good with you?

**Comment by thaJeztah** [comment]
@jfrazelle yes I think it is, but on my phone now. I'll quickly check and merge when I arrive at work

File: api/types/types.go (hunk 11 above)
Comment (thaJeztah) [inline api/types/types.go]: @stevvooe @calavera could one of you create a follow-up pull-request for that?

**Comment by thaJeztah** [comment]
LGTM, thanks @calavera!

**Comment by thaJeztah** [comment]
I added API and changelog impact (changelog - thinking that people may expect the old format)

Do we need impact CLI (for `docker inspect`)?

**Comment by rhatdan** [comment]
Now we need to a patch to change the internal name relabel to Mode, and I will be happy.

**Comment by thaJeztah** [comment]
> Now we need to a patch to change the internal name relabel to Mode, and I will be happy.

@rhatdan I agree (this is confusing). Are you interested to create that PR? (unless there are objections of course)

**Comment by rhatdan** [comment]
Created pull request to modify internal Relabel to Mode

https://github.com/docker/docker/pull/14852
